### PR TITLE
KREST-4591 add topic recreate logic to cluster test harness

### PR DESCRIPTION
**Pull request by ehumber**
Most of the test flakes in this class are due to failures in the setup and the topics not being created.

I can't reproduce this locally unfortunately, but I have put in a pile of retry logic to allow more time, and then have another try at creating the environment.

Also add in a test retry for a flakey test.

### CODE REVIEW

File: kafka-rest/src/test/java/io/confluent/kafkarest/integration/ClusterTestHarness.java [hunk 1]
```diff
@@ -298,11 +307,23 @@ protected Properties getBrokerProperties(int i) {
   @AfterEach
   public void tearDown() throws Exception {
     log.info("Starting teardown of {}", getClass().getSimpleName());
+    tearDownMethod();
+    log.info("Completed teardown of {}", getClass().getSimpleName());
```
Comment (dimitarndimitrov) [inline kafka-rest/src/test/java/io/confluent/kafkarest/integration/ClusterTestHarness.java:311]: nit: You should remove the duplicating log at the end of the new `tearDownMethod()`, like you did in the `setupMethod()`.

File: kafka-rest/src/test/java/io/confluent/kafkarest/integration/ClusterTestHarness.java [hunk 2]
```diff
@@ -298,11 +307,23 @@ protected Properties getBrokerProperties(int i) {
...
+  }
+
+  private void tearDownMethod() throws Exception {
+
+    restProperties.clear();
     if (restServer != null) {
       restServer.stop();
       restServer.join();
     }
```
Comment (dimitarndimitrov) [inline kafka-rest/src/test/java/io/confluent/kafkarest/integration/ClusterTestHarness.java:320]: This has been replaced with `stopRest()` by some of your previous changes, and indeed the PR is conflicting with `master`. Can you please resolve the conflicts?

File: kafka-rest/src/test/java/io/confluent/kafkarest/integration/ClusterTestHarness.java [hunk 3]
```diff
@@ -89,6 +89,7 @@
 public abstract class ClusterTestHarness {
 
   private static final Logger log = LoggerFactory.getLogger(ClusterTestHarness.class);
+  private static final long SLEEP_MS = 500;
```
Comment (dimitarndimitrov) [inline kafka-rest/src/test/java/io/confluent/kafkarest/integration/ClusterTestHarness.java:92]: nit: I think we should rename this and suffix its value with `L`.

File: kafka-rest/src/test/java/io/confluent/kafkarest/integration/ClusterTestHarness.java [hunk 4]
```diff
@@ -332,6 +343,18 @@ protected Properties getBrokerProperties(int i) {
   public void tearDown() throws Exception {
     log.info("Starting teardown of {}", getClass().getSimpleName());
     stopRest();
```
[the remark this answers is not in the document]
Comment (ehumber) [reply kafka-rest/src/test/java/io/confluent/kafkarest/integration/ClusterTestHarness.java:345]: II thought the same, but it's called in two places - once where everything is torn down, and once where just rest is restarted.

File: kafka-rest/src/test/java/io/confluent/kafkarest/integration/ClusterTestHarness.java [hunk 5]
```diff
@@ -442,24 +464,97 @@ protected final int getControllerID() {
...
 
     CreateTopicsResult result =
-        adminClient.createTopics(
-            Collections.singletonList(new NewTopic(topicName, numPartitions, replicationFactor)));
+        createTopicCall(
+            topicName, numPartitions, replicationFactor, replicasAssignments, properties);
 
+    try {
+      result.all().get();
+      throw new InterruptedException();
```
[the remark this answers is not in the document]
Comment (ehumber) [reply kafka-rest/src/test/java/io/confluent/kafkarest/integration/ClusterTestHarness.java:498]: Oh, oops.  Left in from testing.  Good spot.

File: kafka-rest/src/test/java/io/confluent/kafkarest/integration/ClusterTestHarness.java [hunk 6]
```diff
@@ -442,24 +464,97 @@ protected final int getControllerID() {
...
+      if (topicNames.size() == 0) { // Can restart because no topics exist yet
+        log.warn("Restarting the environment as topic creation failed the first time");
+        try {
+          tearDownMethod();
+          pause();
+          setupMethod();
+          pause();
+        } catch (Exception tearDownException) {
+          fail(String.format("Failed to create topic: %s", tearDownException.getMessage()));
+        }
```
[the remark this answers is not in the document]
Comment (ehumber) [reply kafka-rest/src/test/java/io/confluent/kafkarest/integration/ClusterTestHarness.java:511]: The fail logs a message out - it throws an AssertionError with a message in it that is written out in the test.  Or are you looking at a different line to 510?

File: kafka-rest/src/test/java/io/confluent/kafkarest/integration/ClusterTestHarness.java (hunk 1 above)
Comment (ehumber) [reply kafka-rest/src/test/java/io/confluent/kafkarest/integration/ClusterTestHarness.java:347]: I've moved the log start and end lines to the methods doing the actual stopping, as they are called from more than one place - effectively entry and exit logs, which belong in the method doing the work, not the calling method.

I did this for the setup and tear down, and the stop rest call too.

**Comment by ehumber** [comment]
@dimitarndimitrov I thought you'd already reviewed this, but I must have dreamt it :). When you get chance, once you are back in the office, could you take a look please.

**Review by dimitarndimitrov — APPROVED** [review]
+1.

**Pull request merged**